docs/proposals/karmada-operator/api-server-sidecard-containers/README.md

```diff
@@ -0,0 +1,68 @@
+---
+title: Support for API Server Sidecar Containers in Karmada Operator
+authors:
+- "@jabellard"
+reviewers:
+- "@RainbowMango"
+approvers:
+- "@RainbowMango"
+
+creation-date: 2025-02-07
+
+---
+
+# Support for API Server Sidecar Containers in Karmada Operator
+
+## Summary
+
+This proposal aims to enhance the Karmada operator by introducing support for configuring additional containers  for the Karmada API server component.
+By allowing users to define additional containers, this feature unlocks new use cases such as integration with Key Management Service (KMS) plugins for configuring encryption at rest. 
+This enhancement provides greater flexibility while maintaining the declarative nature of the Karmada operator’s deployment model.
+
+## Motivation
+
+Currently, the Karmada operator provisions the API server as a standalone container within a pod. While the API server supports configurability via extra arguments, additional volumes, and volume mounts, it lacks built-in support for defining additional containers.
+By introducing explicit support for additional containers, this feature enables:
+
+- Seamless integration of auxiliary services such as KMS-based encryption providers.
+- Greater deployment flexibility by enabling users to extend the API server’s functionality without modifying the core operator logic.
+
+### Goals
+- Introduce support for configuring sidecar containers for the Karmada API server.
+
+### Non-Goals
+
+- Introduce changes beyond the scope of what is needed to support additional containers for the Karmada API server component.
+
+### API Changes
+
+```go
+// KarmadaAPIServer holds settings for the Karmada API server.
+type KarmadaAPIServer struct {
+    // Other, existing fields omitted for brevity
+    
+    // ExtraContainers specifies a list of additional containers to be deployed
+    // within the Karmada API server pod.
+    // This enables users to integrate auxiliary services such as KMS plugins for configuring encryption at rest.
+    // +optional
+    ExtraContainers []corev1.Container `json:"extraContainers,omitempty"`
+}
+```
+
+### User Stories
+
+#### Story 1
+As an infrastructure engineer, I want to integrate with my organization's KMS to configure encryption at rest for managed Karmada control planes.
+When using a KMS-based encryption provider, a KMS plugin  is required to enable the integration. Once set up, the API server communicates to the plugin over a UNIX domain socket via gRPC. 
+For that to work, when the API server is deployed as a pod as is the case when provisioned by the Karmada operator, the plugin has to run as a sidecar of the API server container.
```

Comment:
Can you add an example of the configuration of `KarmadaAPIServer`?  I guess that makes the user story more straightforward.

Reply:
That's an excellent point. Will add that.

Reply:
Thanks. Appreciated!